homeassistant/components/elkm1/alarm_control_panel.py

```diff
@@ -152,6 +154,21 @@ def _watch_keypad(self, keypad, changeset):
             self._changed_by = username(self._elk, keypad.last_user)
             self.async_write_ha_state()
 
+    def _watch_area(self, area, changeset):
+        if not changeset.get("log_event"):
+            return
+        self._changed_by_keypad = None
+        self._changed_by_id = area.log_number
+        self._changed_by = username(self._elk, area.log_number - 1)
+        self._changed_by_time = "%04d-%02d-%02dT%02d:%02d" % (
```

Comment:
Is this going to be in the same time zone as the `keypad.last_user_time.isoformat()` ?

Reply:
Great question -- as we discussed I think the keypad IC-handling variant should go away in time, but they could be different timezones.  The keypad handling IC code uses the time as recorded when the handler is called and UTC for a timezone.  This means that it's not recording the actual time of the keypad event, but the time when hass got around to processing the event -- yet another undesirable behaviour of using the keypad mechanism for this functionality.

The new code uses the Elk M1 configured timezone (whatever it is) as it's using the actual reported time from the message from the controller, which means it's when Elk recorded the event happening (without hass-induced delay). 

I'm pretty sure we should just get rid of the keypad handler, but per the prior discussion we agreed to keep the old code for num_areas==1 setups. Maybe we change the keypad handler to set attribute keypad_changed_by_time and reserve the changed_by_time for the Elk-report time when LD is on?

I don't see an easy way to make them the same timezone and think the new behaviour is correct, with letting the Elk system report time (you can always figure out the hass time of the message in an automation).

Thoughts?

Reply:
I'm thinking we shouldn't mix data with different time zones in the attributes as that is going to be very confusing.

If we accept that there is always going to be some delay updating the state, is there ever an instance where we the last `LD` response we receive happened in the past?

Reply:
Well, it always was received sometime in the past since it's not a realtime system. 

Keep in mind this mixing is only due the desire to not break for folks with 1 area.  Here are the options I see:

a) for 1 area, use keypad logic exclusively (make it "else add_callback(watch_area)") so you get only the old behaviour.  Consider a breaking change later that requires LD turned on even for 1 area configs and that changes the timezone, too.
b) for > 1 area, use new attributes and leave the old broken attributes around and let that old code run but document those old attributes might not be right when arming/disarming using virtual keypads in HomeAssistant.
c) eliminate the old heuristic that only works correctly when there is 1 area and document that LD needs to be turned on for changed_by* attributes to be added.

I prefer "c" since I don't see the big burden of setting the flag and I view that as the most secure way of interfacing to the alarm system (trusting its time, guaranteeing that we match up with its audit logs), but I'm ok with any of these.

Reply:
I've got quite a few elk systems (all with one area) that I would have to flip that flag which are not all easy to access remotely esp. w/COVID-19 on so I'm not a big fan of a breaking change right now 🙈 

I'm digging for a way to get the timezone of the elk system so we can convert.

Reply:
It looks like `RR` will tell us the time now but its not particularly portable as the only timezone data is an `S` field that tells `S - Daylight Savings Time: “0”=Not active, “1”= Active.`

Reply:
With one area `changed_by_time: '2020-05-28T18:43'`

Reply:
With LD turned on `'2020-05-28T18:47'`

Reply:
With keypad code commented out  `'2020-05-28T18:55'`

Reply:
```
# rm -fv /etc/localtime
```
Trying with localtime deleted on the hass docker container

Reply:
With localtime removed and keypad code commented out
``` '2020-05-28T18:57'```

Reply:
So with either method as long as the timezone between HASS and elkm1 agree it will be in localtime.

Reply:
Since this is a local integration it would be a misconfiguration if the timezones between hass and elk didn't match (excluding the odd vpn across timezone which is safe it say is not going to be supported).

Reply:
I'm comfortable with it as-is.

Reply:
Time data in state attribute should always be absolute UTC.

Please use f-strings instead of old style string formatting.